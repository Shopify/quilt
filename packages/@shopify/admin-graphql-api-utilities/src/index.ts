const GID_REGEXP = /\/(\w+(-\w+)*)$/;

export function parseGid(gid: string): string {
  // appends forward slash to help identify invalid id
  const id = `/${gid}`;
  const matches = id.match(GID_REGEXP);
  if (matches && matches[1] !== undefined) {
    return matches[1];
  }
  throw new Error(`Invalid gid: ${gid}`);
}

export function composeGid(key: string, id: number | string): string {
  return `gid://shopify/${key}/${id}`;
}

interface Edge<T> {
  node: T;
}

export function nodesFromEdges<T>(edges: Edge<T>[]): T[] {
  return edges.map(({node}) => node);
}

export function keyFromEdges<T, K extends keyof T>(
  edges: Edge<T>[],
  key: K,
): T[K][] {
  return edges.map(({node}) => node[key]);
}
