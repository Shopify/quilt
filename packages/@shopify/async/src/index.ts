export * from './types';
export * from './resolver';
