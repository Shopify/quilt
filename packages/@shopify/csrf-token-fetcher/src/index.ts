const CSRF_SELECTOR = 'meta[name=csrf-token]';
let token: string;

export default function getCSRFToken(scope = document) {
  if (token) {
    return token;
  }
  const tokenNode = scope.querySelector(CSRF_SELECTOR);
  if (tokenNode) {
    const content = tokenNode.getAttribute('content');
    if (content) {
      token = content;
      return token;
    }
  }
  return '';
}

export function clearCSRFToken() {
  token = '';
}
