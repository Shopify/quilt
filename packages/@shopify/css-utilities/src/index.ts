import classNames from 'classnames';

export {classNames};
export {variationName} from './variation-name';
