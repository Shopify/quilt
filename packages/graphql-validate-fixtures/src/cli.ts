/* eslint no-console: off */

import {relative, basename} from 'path';

import yargs from 'yargs';
import * as glob from 'glob';
import chalk from 'chalk';

import {evaluateFixtures} from '.';

const argv = yargs
  .usage('Usage: $0 <fixtures> [options]')
  .option('cwd', {
    required: false,
    default: process.cwd(),
    normalize: true,
    type: 'string',
    describe: 'Working directory where the .graphqlconfig is located',
  })
  .option('show-passes', {
    type: 'boolean',
    default: false,
    describe:
      'Display passing fixtures along with failures in the console output',
  })
  .help().argv;

evaluateFixtures(glob.sync(argv._[0]), {
  cwd: argv.cwd,
})
  .then((evaluations) => {
    let passed = 0;
    let failed = 0;
    let lastFailed = false;

    console.log();

    evaluations.forEach((evaluation) => {
      const relativePath = relative(argv.cwd, evaluation.fixturePath);
      const formattedPath = `${chalk.dim(
        relativePath.replace(basename(relativePath), ''),
      )}${chalk.bold(basename(relativePath))}`;

      if (evaluation.scriptError) {
        failed += 1;
        lastFailed = true;
        console.log(`${chalk.inverse.bold.red(' FAIL ')} ${formattedPath}`);
        console.log(evaluation.scriptError.message);
        if (evaluation.scriptError.stack) {
          console.log(
            chalk.dim(
              evaluation.scriptError.stack
                .replace(evaluation.scriptError.message, '')
                .replace(/^Error:\s*\n/, ''),
            ),
          );
        }
        console.log();
      } else if (evaluation.validationErrors.length === 0) {
        passed += 1;
        lastFailed = false;
        if (argv.showPasses) {
          console.log(`${chalk.inverse.bold.green(' PASS ')} ${formattedPath}`);
        }
      } else {
        failed += 1;
        lastFailed = true;

        console.log(`${chalk.inverse.bold.red(' FAIL ')} ${formattedPath}`);
        evaluation.validationErrors.forEach((error) => {
          console.log(`${error.keyPath} ${chalk.red(error.message)}`);
        });
        console.log();
      }
    });

    if (passed + failed === 0) {
      console.log(chalk.bold.yellow('No tests run'));
      return;
    }

    if (!lastFailed) {
      console.log();
    }

    if (passed > 0) {
      console.log(chalk.bold.green(`${passed} passed`));
    }

    if (failed > 0) {
      console.log(chalk.bold.red(`${failed} failed`));
      process.exit(1);
    }
  })
  .catch((error) => {
    console.log();
    console.log(`${chalk.inverse.bold.red(' ERROR ')} ${error.message}`);
    console.log(
      chalk.dim(
        error.stack.replace(error.message, '').replace(/^Error:\s*\n/, ''),
      ),
    );
    process.exit(1);
  });
