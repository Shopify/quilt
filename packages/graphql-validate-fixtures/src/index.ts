import {resolve} from 'path';

import {readFile, readJSON} from 'fs-extra';
import {Source, parse, concatAST, GraphQLSchema} from 'graphql';
import {getGraphQLConfig, GraphQLProjectConfig} from 'graphql-config';
import {
  getGraphQLProjectIncludedFilePaths,
  getGraphQLProjects,
} from 'graphql-config-utilities';
import {compile} from 'graphql-tool-utilities';

import {
  Fixture,
  Validation,
  validateFixture,
  getOperationForFixture,
  MissingOperationError,
  AmbiguousOperationNameError,
  GraphQLProjectAST,
} from './validate';

export interface Options {
  cwd: string;
}

export interface Evaluation extends Validation {
  scriptError?: Error;
}

export async function evaluateFixtures(
  fixturePaths: string[],
  {cwd}: Options,
): Promise<Evaluation[]> {
  const config = getGraphQLConfig(resolve(cwd));

  const projectASTCollection = await Promise.all(
    getGraphQLProjects(config).map(getOperationsForProject),
  );

  return runForEachFixture(fixturePaths, fixture =>
    evaluateFixture(fixture, projectASTCollection),
  );
}

async function getOperationsForProject(
  projectConfig: GraphQLProjectConfig,
): Promise<GraphQLProjectAST> {
  const operationPaths = await getGraphQLProjectIncludedFilePaths(
    projectConfig,
  );

  const operationSources = await Promise.all(
    operationPaths.map(loadOperationSource),
  );

  const document = concatAST(
    operationSources.map((source: Source) => {
      try {
        return parse(source);
      } catch (error) {
        throw new Error(
          `Error parsing '${source.name}':\n\n${error.message.replace(
            /Syntax Error.*?\(.*?\) /,
            '',
          )}`,
        );
      }
    }),
  );

  let schema: GraphQLSchema;

  try {
    schema = projectConfig.getSchema();
  } catch (error) {
    throw new Error(
      `Error parsing '${projectConfig.schemaPath}':\n\n${error.message.replace(
        /Syntax Error.*?\(.*?\) /,
        '',
      )}`,
    );
  }

  return {
    ast: compile(schema, document),
    config: projectConfig,
  };
}

async function loadOperationSource(filePath: string) {
  const body = await readFile(filePath, 'utf8');

  return new Source(body, filePath);
}

function runForEachFixture<T extends Partial<Evaluation>>(
  fixturePaths: string[],
  runner: (fixture: Fixture) => T,
): Promise<Evaluation[]> {
  return Promise.all(
    fixturePaths.map(async fixturePath => {
      try {
        const fixture = await readJSON(fixturePath);
        return {
          fixturePath,
          ...(runner({path: fixturePath, content: fixture}) as any),
        };
      } catch (error) {
        return {
          fixturePath,
          scriptError: error,
          validationErrors: [],
        };
      }
    }),
  );
}

function evaluateFixture(
  fixture: Fixture,
  projectASTCollection: GraphQLProjectAST[],
): Evaluation {
  try {
    const {
      operation,
      projectAST: {ast},
    } = getOperationForFixture(fixture, projectASTCollection);

    return validateFixture(fixture, ast, operation);
  } catch (error) {
    if (
      error instanceof MissingOperationError ||
      error instanceof AmbiguousOperationNameError
    ) {
      return {
        fixturePath: fixture.path,
        validationErrors: [error],
      };
    }

    throw error;
  }
}
