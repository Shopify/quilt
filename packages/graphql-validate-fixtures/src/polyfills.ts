// eslint-disable-next-line no-warning-comments
// TODO: Remove once we kill NodeJs 10 support
import 'core-js/features/array/flat-map';
