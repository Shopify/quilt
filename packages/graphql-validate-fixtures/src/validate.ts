import {dirname, basename} from 'path';

import {
  GraphQLType,
  GraphQLNonNull,
  GraphQLList,
  GraphQLString,
  GraphQLInt,
  GraphQLFloat,
  GraphQLBoolean,
  GraphQLLeafType,
  isEnumType,
  isListType,
  isNonNullType,
  isObjectType,
  isScalarType,
} from 'graphql';
import {GraphQLProjectConfig} from 'graphql-config';
import {resolveProjectName} from 'graphql-config-utilities';
import {AST, Field, Operation} from 'graphql-tool-utilities';

export type KeyPath = string;

export interface Fixture {
  path: string;
  content: any;
}

export interface Error {
  keyPath?: KeyPath;
  message: string;
}

export interface GraphQLProjectAST {
  config: GraphQLProjectConfig;
  ast: AST;
}

export interface FoundOperation {
  lookForOperationName: string;
  operation: Operation;
  projectAST: GraphQLProjectAST;
}

export class MissingOperationError extends Error {
  constructor(
    {path}: Fixture,
    lookForOperationNames: string[],
    projectASTCollection: GraphQLProjectAST[],
  ) {
    super(
      [
        `Could not find a matching operation for '${path}'`,
        `(looked for ${lookForOperationNames.join(', ')}).`,
        `Make sure to put your fixture in a folder named the same as the operation,`,
        `or add an '${OPERATION_MARKER}' key indicating the operation.`,
        `Available operations: ${projectASTCollection
          .flatMap(({ast: {operations}}) => Object.keys(operations))
          .join(', ')}`,
      ].join(' '),
    );
  }
}

export class AmbiguousOperationNameError extends Error {
  constructor({path}: Fixture, foundOperations: FoundOperation[]) {
    super(
      [
        `Ambiguous operation name found for '${path}'`,
        `(found ${Array.from(
          new Set(
            foundOperations.map(
              ({lookForOperationName}) => lookForOperationName,
            ),
          ),
        ).join(', ')})`,
        `in projects:`,
        `${foundOperations
          .map(({projectAST: {config}}) => resolveProjectName(config))
          .join(', ')}.`,
        `Try renaming the operation in one of the projects listed and updating`,
        `the fixture folder name or use an '${OPERATION_MARKER}' key indicating`,
        `the new operation name.`,
      ].join(' '),
    );
  }
}

export interface Validation {
  fixturePath: string;
  operationName?: string;
  operationType?: string;
  operationPath?: string;
  validationErrors: Error[];
}

const OPERATION_MARKER = '@operation';

function normalizeOperationName(operationName: string): string;
function normalizeOperationName(
  operationName: string | undefined,
): string | undefined;
function normalizeOperationName(operationName: string | undefined) {
  return operationName
    ? operationName.replace(/(Query|Mutation|Subscription)$/i, '')
    : undefined;
}

export function getOperationNames(fixture: Fixture): string[] {
  const fixtureDirectoryName = basename(dirname(fixture.path));
  const operationMarkerName: string | undefined =
    fixture.content[OPERATION_MARKER];

  return Array.from(
    new Set([
      fixtureDirectoryName,
      operationMarkerName,
      normalizeOperationName(fixtureDirectoryName),
      normalizeOperationName(operationMarkerName),
    ]),
  ).filter(
    (operationName): operationName is string =>
      typeof operationName === 'string',
  );
}

export function findOperations(
  lookForOperationNames: string[],
  projectASTCollection: GraphQLProjectAST[],
) {
  return projectASTCollection
    .map<FoundOperation | null>((projectAST) => {
      for (const lookForOperationName of lookForOperationNames) {
        const operation = projectAST.ast.operations[lookForOperationName];

        if (operation) {
          return {
            lookForOperationName,
            operation,
            projectAST,
          };
        }
      }

      return null;
    })
    .filter((match): match is FoundOperation => Boolean(match));
}

export function getOperationForFixture(
  fixture: Fixture,
  projectASTCollection: GraphQLProjectAST[],
) {
  const lookForOperationNames = getOperationNames(fixture);
  const operations = findOperations(
    lookForOperationNames,
    projectASTCollection,
  );

  if (operations.length === 0) {
    throw new MissingOperationError(
      fixture,
      lookForOperationNames,
      projectASTCollection,
    );
  }

  if (operations.length > 1) {
    throw new AmbiguousOperationNameError(fixture, operations);
  }

  return operations[0];
}

export interface FixtureOperation {
  fixture: Fixture;
  operation: Operation;
  operationName: string;
}

export function validateFixture(
  fixture: Fixture,
  ast: AST,
  operation: Operation,
): Validation {
  const {fields = [], filePath, operationType} = operation;
  const value = {...fixture.content};
  delete value[OPERATION_MARKER];

  return {
    fixturePath: fixture.path,
    operationName: operation.operationName,
    operationType,
    operationPath: filePath === 'GraphQL request' ? undefined : filePath,
    validationErrors: fields.reduce((allErrors: Error[], field) => {
      return allErrors.concat(
        validateValueAgainstFieldDescription(
          value[field.responseName],
          field,
          '',
          ast,
        ),
      );
    }, []),
  };
}

function validateValueAgainstFieldDescription(
  value: any,
  fieldDescription: Field,
  parentKeyPath: string,
  ast: AST,
): Error[] {
  const {type, responseName} = fieldDescription;
  const keyPath = updateKeyPath(parentKeyPath, responseName);
  const typeErrors = validateValueAgainstType(value, type, keyPath, {
    shallow: true,
  });

  if (typeErrors.length > 0) {
    return typeErrors;
  }

  return Array.isArray(value)
    ? validateListAgainstFieldDescription(
        value,
        fieldDescription,
        type as GraphQLList<GraphQLType>,
        keyPath,
        ast,
      )
    : validateValueAgainstObjectFieldDescription(
        value,
        fieldDescription,
        keyPath,
        ast,
      );
}

function validateValueAgainstObjectFieldDescription(
  value: any,
  fieldDescription: Field,
  keyPath: string,
  ast: AST,
) {
  if (value == null || typeof value !== 'object' || Array.isArray(value)) {
    return [];
  }

  const {fields = [], fragmentSpreads = [], type} = fieldDescription;

  const fragmentFields: Field[] = [];

  if (fragmentSpreads) {
    fragmentSpreads
      .map((spread) => ast.fragments[spread])
      .forEach((fragment) => {
        fragment.fields.forEach((field) => {
          if (fields.some(({fieldName}) => fieldName === field.fieldName)) {
            return;
          }

          const isGuaranteedTypeMatch = fragment.possibleTypes.includes(
            makeTypeNullable(type),
          );

          fragmentFields.push(
            isGuaranteedTypeMatch
              ? field
              : {...field, type: makeTypeNullable(field.type)},
          );
        });
      });
  }

  return validateValueAgainstFields(
    value,
    fields.concat(fragmentFields),
    keyPath,
    ast,
  );
}

function makeTypeNullable(type: GraphQLType) {
  return isNonNullType(type) ? type.ofType : type;
}

function validateListAgainstFieldDescription(
  value: any[],
  fieldDescription: Field,
  type: GraphQLList<GraphQLType>,
  keyPath: string,
  ast: AST,
): Error[] {
  const itemType = isNonNullType(type)
    ? (type as GraphQLNonNull<GraphQLList<GraphQLType>>).ofType.ofType
    : (type as GraphQLList<GraphQLType>).ofType;

  return value.reduce((allErrors, item, index) => {
    const itemKeyPath = updateKeyPath(keyPath, index);
    const itemTypeErrors = validateValueAgainstType(
      item,
      itemType,
      itemKeyPath,
      {shallow: true},
    );

    if (itemTypeErrors.length > 0) {
      return allErrors.concat(itemTypeErrors);
    }

    return Array.isArray(item)
      ? allErrors.concat(
          validateListAgainstFieldDescription(
            item,
            fieldDescription,
            itemType as GraphQLList<GraphQLType>,
            itemKeyPath,
            ast,
          ),
        )
      : allErrors.concat(
          validateValueAgainstObjectFieldDescription(
            item,
            fieldDescription,
            itemKeyPath,
            ast,
          ),
        );
  }, []);
}

function validateValueAgainstFields(
  value: {[key: string]: any},
  fields: Field[],
  keyPath: KeyPath,
  ast: AST,
) {
  const finalValue = value || {};

  const excessFields = Object.keys(finalValue)
    .filter((key) => fields.every(({responseName}) => key !== responseName))
    .map((key) => error(keyPath, `has key '${key}' not present in the query`));

  return fields.reduce((allErrors, field) => {
    return allErrors.concat(
      validateValueAgainstFieldDescription(
        finalValue[field.responseName],
        field,
        keyPath,
        ast,
      ),
    );
  }, excessFields);
}

interface TypeValidationOptions {
  shallow: boolean;
}

function validateValueAgainstType(
  value: any,
  type: GraphQLType,
  keyPath: KeyPath,
  options: TypeValidationOptions = {shallow: false},
): Error[] {
  const {shallow} = options;

  if (isNonNullType(type)) {
    return value == null
      ? [error(keyPath, `should be non-null but was ${String(value)}`)]
      : validateValueAgainstType(value, type.ofType, keyPath, options);
  }

  if (value === null) {
    return [];
  }

  const valueType = typeof value;

  if (isListType(type)) {
    if (!Array.isArray(value)) {
      return [
        error(
          keyPath,
          `should be an array (or null), but was ${articleForType(
            valueType,
          )} ${valueType}`,
        ),
      ];
    }

    return shallow
      ? []
      : value.reduce(
          (allErrors: Error[], item, index) =>
            allErrors.concat(
              validateValueAgainstType(
                item,
                type.ofType,
                updateKeyPath(keyPath, index),
                options,
              ),
            ),
          [],
        );
  }

  if (value === undefined) {
    const typeName = nameForType(type as GraphQLLeafType);
    return [
      error(
        keyPath,
        `should be ${articleForType(
          typeName,
        )} ${typeName} (or null), but was undefined`,
      ),
    ];
  }

  if (isObjectType(type)) {
    if (valueType === 'object') {
      if (shallow) {
        return [];
      } else {
        const fields = type.getFields();
        return Object.keys(value).reduce((fieldErrors: Error[], key) => {
          const fieldKeyPath = updateKeyPath(keyPath, key);

          return fields[key] == null
            ? fieldErrors.concat([
                error(
                  fieldKeyPath,
                  `does not exist on type ${
                    type.name
                  } (available fields: ${Object.keys(fields).join(', ')})`,
                ),
              ])
            : fieldErrors.concat(
                validateValueAgainstType(
                  value[key],
                  fields[key].type,
                  fieldKeyPath,
                  options,
                ),
              );
        }, []);
      }
    } else {
      return [error(keyPath, `should be an object but was a ${valueType}`)];
    }
  }

  if (type === GraphQLString) {
    return valueType === 'string'
      ? []
      : [error(keyPath, `should be a string but was a ${valueType}`)];
  } else if (type === GraphQLInt) {
    if (typeof value === 'number') {
      return Number.isInteger(value)
        ? []
        : [error(keyPath, 'should be an integer but was a float')];
    }

    return [error(keyPath, `should be an integer but was a ${valueType}`)];
  } else if (type === GraphQLFloat) {
    return Number.isNaN(value)
      ? [error(keyPath, `should be a float but was a ${valueType}`)]
      : [];
  } else if (type === GraphQLBoolean) {
    return typeof value === 'boolean'
      ? []
      : [error(keyPath, `should be a boolean but was a ${valueType}`)];
  }

  if (isScalarType(type)) {
    return type.parseValue(value) == null
      ? [error(keyPath, `value does not match scalar ${nameForType(type)}`)]
      : [];
  }

  if (isEnumType(type)) {
    return type.parseValue(value) == null
      ? [
          error(
            keyPath,
            `value does not match enum ${nameForType(
              type,
            )} (available values: ${type
              .getValues()
              .map((enumValue) => enumValue.value)
              .join(', ')})`,
          ),
        ]
      : [];
  }

  return [];
}

const CUSTOM_NAMES = {
  [GraphQLBoolean.name]: 'boolean',
  [GraphQLFloat.name]: 'float',
  [GraphQLInt.name]: 'integer',
  [GraphQLString.name]: 'string',
};

function nameForType(type: GraphQLLeafType) {
  return Object.prototype.hasOwnProperty.call(CUSTOM_NAMES, type.name)
    ? CUSTOM_NAMES[type.name]
    : type.name;
}

const TYPES_WITH_ARTICLE_AN = ['object', 'integer', 'array'];

function articleForType(type: string) {
  return TYPES_WITH_ARTICLE_AN.includes(type) ? 'an' : 'a';
}

function updateKeyPath(keyPath: KeyPath, newKey: string | number) {
  if (typeof newKey === 'number') {
    return `${keyPath}[${newKey}]`;
  }

  return keyPath ? `${keyPath}.${newKey}` : newKey;
}

function error(keyPath: KeyPath, message: string): Error {
  return {keyPath, message};
}
