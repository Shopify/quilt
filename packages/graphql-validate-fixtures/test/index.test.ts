/* eslint-disable @shopify/jest/no-snapshots */
import {join} from 'path';

import * as glob from 'glob';

import {stripFullFilePaths} from '../../../tests/utilities';
import {evaluateFixtures, Options} from '../src/index';

const rootFixtureDirectory = join(__dirname, 'fixtures');

describe('evaluateFixtures()', () => {
  it('handles fixtures without errors', async () => {
    expect(await evaluateFixturesForFixturePath('all-clear')).toMatchSnapshot();
  });

  it('handles multi-project fixtures without errors', async () => {
    expect(
      await evaluateFixturesForFixturePath('multi-project'),
    ).toMatchSnapshot();
  });

  it('handles ambiguous operation names in multi-project fixtures with errors', async () => {
    expect(
      await evaluateFixturesForFixturePath('ambiguous-operation'),
    ).toMatchSnapshot();
  });

  it('handles missing operation names in multi-project fixtures with errors', async () => {
    expect(
      await evaluateFixturesForFixturePath('missing-operation'),
    ).toMatchSnapshot();
  });

  it('handles fixtures with errors', async () => {
    expect(
      await evaluateFixturesForFixturePath('fixture-errors'),
    ).toMatchSnapshot();
  });

  it('handles fixtures that are invalid json', async () => {
    expect(
      await evaluateFixturesForFixturePath('fixture-invalid'),
    ).toMatchSnapshot();
  });

  it('throws an error when the schema is not found', async () => {
    await expect(
      evaluateFixturesForFixturePath('missing-schema'),
    ).rejects.toMatchSnapshot();
  });

  it('throws an error when there are malformed GraphQL documents', async () => {
    await expect(
      evaluateFixturesForFixturePath('malformed-query'),
    ).rejects.toMatchSnapshot();
  });
});

async function evaluateFixturesForFixturePath(
  fixture: string,
  options?: Partial<Options>,
) {
  try {
    const result = await evaluateFixtures(detailsForFixture(fixture), {
      cwd: join(rootFixtureDirectory, fixture),
      ...options,
    });
    return stripFullFilePaths(result);
  } catch (error) {
    throw stripFullFilePaths(error);
  }
}

function detailsForFixture(fixture: string) {
  return glob.sync(join(rootFixtureDirectory, fixture, 'fixtures/**/*.json'));
}

/* eslint-enable @shopify/jest/no-snapshots */
