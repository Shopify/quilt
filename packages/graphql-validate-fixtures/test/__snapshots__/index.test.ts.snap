// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`evaluateFixtures() handles ambiguous operation names in multi-project fixtures with errors 1`] = `
Array [
  Object {
    "fixturePath": "packages/graphql-validate-fixtures/test/fixtures/ambiguous-operation/fixtures/projectA/AmbiguousQuery/fixture.json",
    "validationErrors": Array [
      [Error: Ambiguous operation name found for 'packages/graphql-validate-fixtures/test/fixtures/ambiguous-operation/fixtures/projectA/AmbiguousQuery/fixture.json' (found Ambiguous) in projects: projectA, projectB. Try renaming the operation in one of the projects listed and updating the fixture folder name or use an '@operation' key indicating the new operation name.],
    ],
  },
  Object {
    "fixturePath": "packages/graphql-validate-fixtures/test/fixtures/ambiguous-operation/fixtures/projectB/AmbiguousQuery/fixture.json",
    "validationErrors": Array [
      [Error: Ambiguous operation name found for 'packages/graphql-validate-fixtures/test/fixtures/ambiguous-operation/fixtures/projectB/AmbiguousQuery/fixture.json' (found Ambiguous) in projects: projectA, projectB. Try renaming the operation in one of the projects listed and updating the fixture folder name or use an '@operation' key indicating the new operation name.],
    ],
  },
]
`;

exports[`evaluateFixtures() handles fixtures that are invalid json 1`] = `
Array [
  Object {
    "fixturePath": "packages/graphql-validate-fixtures/test/fixtures/fixture-invalid/fixtures/another-fixture.json",
    "scriptError": [SyntaxError: packages/graphql-validate-fixtures/test/fixtures/fixture-invalid/fixtures/another-fixture.json: Unexpected string in JSON at position 35],
    "validationErrors": Array [],
  },
  Object {
    "fixturePath": "packages/graphql-validate-fixtures/test/fixtures/fixture-invalid/fixtures/MyQuery/fixture.json",
    "operationName": "My",
    "operationPath": "packages/graphql-validate-fixtures/test/fixtures/fixture-invalid/queries/Query.graphql",
    "operationType": "query",
    "validationErrors": Array [],
  },
]
`;

exports[`evaluateFixtures() handles fixtures with errors 1`] = `
Array [
  Object {
    "fixturePath": "packages/graphql-validate-fixtures/test/fixtures/fixture-errors/fixtures/another-fixture.json",
    "operationName": "Another",
    "operationPath": "packages/graphql-validate-fixtures/test/fixtures/fixture-errors/queries/Another.graphql",
    "operationType": "query",
    "validationErrors": Array [
      Object {
        "keyPath": "name",
        "message": "should be a string but was a number",
      },
      Object {
        "keyPath": "age",
        "message": "should be non-null but was undefined",
      },
    ],
  },
  Object {
    "fixturePath": "packages/graphql-validate-fixtures/test/fixtures/fixture-errors/fixtures/MyQuery/fixture.json",
    "operationName": "My",
    "operationPath": "packages/graphql-validate-fixtures/test/fixtures/fixture-errors/queries/Query.graphql",
    "operationType": "query",
    "validationErrors": Array [],
  },
]
`;

exports[`evaluateFixtures() handles fixtures without errors 1`] = `
Array [
  Object {
    "fixturePath": "packages/graphql-validate-fixtures/test/fixtures/all-clear/fixtures/another-another-fixture.json",
    "operationName": "Another",
    "operationPath": "packages/graphql-validate-fixtures/test/fixtures/all-clear/queries/Another.graphql",
    "operationType": "query",
    "validationErrors": Array [],
  },
  Object {
    "fixturePath": "packages/graphql-validate-fixtures/test/fixtures/all-clear/fixtures/another-fixture.json",
    "operationName": "Another",
    "operationPath": "packages/graphql-validate-fixtures/test/fixtures/all-clear/queries/Another.graphql",
    "operationType": "query",
    "validationErrors": Array [],
  },
  Object {
    "fixturePath": "packages/graphql-validate-fixtures/test/fixtures/all-clear/fixtures/MyQuery/fixture.json",
    "operationName": "My",
    "operationPath": "packages/graphql-validate-fixtures/test/fixtures/all-clear/queries/Query.graphql",
    "operationType": "query",
    "validationErrors": Array [],
  },
]
`;

exports[`evaluateFixtures() handles missing operation names in multi-project fixtures with errors 1`] = `
Array [
  Object {
    "fixturePath": "packages/graphql-validate-fixtures/test/fixtures/missing-operation/fixtures/projectA/AQuery/fixture.json",
    "operationName": "A",
    "operationPath": "packages/graphql-validate-fixtures/test/fixtures/missing-operation/queries/projectA/Query.graphql",
    "operationType": "query",
    "validationErrors": Array [],
  },
  Object {
    "fixturePath": "packages/graphql-validate-fixtures/test/fixtures/missing-operation/fixtures/projectB/Missing/fixture.json",
    "validationErrors": Array [
      [Error: Could not find a matching operation for 'packages/graphql-validate-fixtures/test/fixtures/missing-operation/fixtures/projectB/Missing/fixture.json' (looked for Missing). Make sure to put your fixture in a folder named the same as the operation, or add an '@operation' key indicating the operation. Available operations: A, B],
    ],
  },
]
`;

exports[`evaluateFixtures() handles multi-project fixtures without errors 1`] = `
Array [
  Object {
    "fixturePath": "packages/graphql-validate-fixtures/test/fixtures/multi-project/fixtures/projectA/AQuery/fixture.json",
    "operationName": "A",
    "operationPath": "packages/graphql-validate-fixtures/test/fixtures/multi-project/queries/projectA/Query.graphql",
    "operationType": "query",
    "validationErrors": Array [],
  },
  Object {
    "fixturePath": "packages/graphql-validate-fixtures/test/fixtures/multi-project/fixtures/projectB/BQuery/fixture.json",
    "operationName": "B",
    "operationPath": "packages/graphql-validate-fixtures/test/fixtures/multi-project/queries/projectB/Query.graphql",
    "operationType": "query",
    "validationErrors": Array [],
  },
]
`;

exports[`evaluateFixtures() throws an error when the schema is not found 1`] = `
[Error: Error parsing 'schema.json':


      Unable to find any GraphQL type definitions for the following pointers:
        
          - schema.json
          ]
`;

exports[`evaluateFixtures() throws an error when there are malformed GraphQL documents 1`] = `
[Error: Error parsing 'packages/graphql-validate-fixtures/test/fixtures/malformed-query/queries/Another.graphql':

Syntax Error: Expected {, found Name "name"]
`;
